floors/rule.go

```diff
@@ -0,0 +1,382 @@
+		return CATCH_ALL
+	}
+
+	if len(bidRequest.Ext) > 0 {
+		err := json.Unmarshal(bidRequest.Ext, &requestExt)
+		if err != nil {
+			return CATCH_ALL
+		}
+	}
+
+	if requestExt.Prebid.Channel != nil {
+		return requestExt.Prebid.Channel.Name
+	}
+	return CATCH_ALL
+}
+
+func getpbadslot(imp openrtb2.Imp) string {
+	value := CATCH_ALL
+	pbAdSlot, err := jsonparser.GetString(imp.Ext, "data", "pbadslot")
+	if err == nil {
+		value = pbAdSlot
+	}
+	return value
+}
+
+func isMobileDevice(userAgent string) bool {
+	isMobile, err := regexp.MatchString("(?i)Phone|iPhone|Android|Mobile", userAgent)
+	if err != nil {
+		return false
+	}
+	return isMobile
+}
+
+func isTabletDevice(userAgent string) bool {
+	isTablet, err := regexp.MatchString("(?i)tablet|iPad|Windows NT", userAgent)
+	if err != nil {
+		return false
+	}
+	return isTablet
+}
+
+func prepareRuleCombinations(keys []string, numSchemaFields int, delimiter string) []string {
+	var subset []string
+	var comb []int
+	var desiredkeys [][]string
+	var ruleKeys []string
+
+	segNum := 1 << numSchemaFields
+	for i := 0; i < numSchemaFields; i++ {
+		subset = append(subset, strings.ToLower(keys[i]))
+		comb = append(comb, i)
+	}
+	desiredkeys = append(desiredkeys, subset)
+	for numWildCart := 1; numWildCart <= numSchemaFields; numWildCart++ {
+		newComb := GenerateCombinations(comb, numWildCart, segNum)
+		for i := 0; i < len(newComb); i++ {
+			eachSet := make([]string, len(desiredkeys[0]))
+			_ = copy(eachSet, desiredkeys[0])
+			for j := 0; j < len(newComb[i]); j++ {
+				eachSet[newComb[i][j]] = CATCH_ALL
+			}
+			desiredkeys = append(desiredkeys, eachSet)
+		}
+	}
+	ruleKeys = PrepareRuleKeys(desiredkeys, delimiter)
+	return ruleKeys
+}
+
+func PrepareRuleKeys(desiredkeys [][]string, delimiter string) []string {
+	var ruleKeys []string
+	for i := 0; i < len(desiredkeys); i++ {
+		subset := desiredkeys[i][0]
+		for j := 1; j < len(desiredkeys[i]); j++ {
+			subset += delimiter + desiredkeys[i][j]
+		}
+		ruleKeys = append(ruleKeys, subset)
+	}
+	return ruleKeys
+}
+
+func GenerateCombinations(set []int, numWildCart int, segNum int) (comb [][]int) {
+	length := uint(len(set))
+
+	if numWildCart > len(set) {
+		numWildCart = len(set)
+	}
+
+	for subsetBits := 1; subsetBits < (1 << length); subsetBits++ {
+		if numWildCart > 0 && bits.OnesCount(uint(subsetBits)) != numWildCart {
+			continue
+		}
+		var subset []int
+		for object := uint(0); object < length; object++ {
+			if (subsetBits>>object)&1 == 1 {
+				subset = append(subset, set[object])
+			}
+		}
+		comb = append(comb, subset)
+	}
+
+	// Sort combinations based on priority mentioned in https://docs.prebid.org/dev-docs/modules/floors.html#rule-selection-process
+	sort.SliceStable(comb, func(i, j int) bool {
+		wt1 := 0
+		for k := 0; k < len(comb[i]); k++ {
+			wt1 += 1 << (segNum - comb[i][k])
+		}
+
+		wt2 := 0
+		for k := 0; k < len(comb[j]); k++ {
+			wt2 += 1 << (segNum - comb[j][k])
+		}
+		return wt1 < wt2
+	})
```

Comment:
Consider extracting the sort algorithm and testing it separately.

Reply:
This is getting tested in TestPrepareRuleCombinations(),
Let us know if it serves purpose.

Reply:
@bsardo what is your opinion?